File: Java/Double Circular/DoubleCircular.java
class DoublyCircularNode {
    int data;
    DoublyCircularNode next;
    DoublyCircularNode prev;

    DoublyCircularNode(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

class DoublyCircularLinkedList {
    DoublyCircularNode head;

    public DoublyCircularLinkedList() {
        this.head = null;
    }

    public void insert(int data) {
        DoublyCircularNode newNode = new DoublyCircularNode(data);
        if (head == null) {
            head = newNode;
            newNode.next = newNode;
            newNode.prev = newNode;
        } else {
            DoublyCircularNode last = head.prev;
            last.next = newNode;
            newNode.next = head;
            newNode.prev = last;
            head.prev = newNode;
        }
    }

    public void deleteNode(int key) {
        if (head == null) return;

        DoublyCircularNode current = head;

        // If the head node holds the key
        if (current.data == key) {
            if (current.next == head) {
                head = null;
                return;
            }

            DoublyCircularNode last = head.prev;
            last.next = current.next;
            current.next.prev = last;
            head = current.next;
            return;
        }

        // Search for the node to be deleted
        while (current.next != head && current.data != key) {
            current = current.next;
        }

        if (current.data == key) {
            current.prev.next = current.next;
            current.next.prev = current.prev;
        }
    }

    public void display() {
        if (head == null) return;
        DoublyCircularNode current = head;
        do {
            System.out.print(current.data + " <-> ");
            current = current.next;
        } while (current != head);
        System.out.println("(head)");
    }

    public static void main(String[] args) {
        DoublyCircularLinkedList list = new DoublyCircularLinkedList();
        list.insert(3);
        list.insert(2);
        list.insert(1);

        System.out.println("Original List:");
        list.display();

        list.deleteNode(2);
        System.out.println("After Deleting 2:");
        list.display();
    }
}

File: Python/Double Circular/DoubleCircular.py
class DoublyCircularNode:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyCircularLinkedList:
    def __init__(self):
        self.head = None

    def insert(self, data):
        new_node = DoublyCircularNode(data)
        if self.head is None:
            self.head = new_node
            new_node.next = new_node
            new_node.prev = new_node
        else:
            last = self.head.prev
            last.next = new_node
            new_node.next = self.head
            new_node.prev = last
            self.head.prev = new_node

    def delete_node(self, key):
        if self.head is None:
            return

        current = self.head

        # If the head node holds the key
        if current.data == key:
            if current.next == self.head:
                self.head = None
                return

            last = self.head.prev
            last.next = current.next
            current.next.prev = last
            self.head = current.next
            return

        # Search for the node to be deleted
        while current.next != self.head and current.data != key:
            current = current.next

        if current.data == key:
            current.prev.next = current.next
            current.next.prev = current.prev

    def display(self):
        if self.head is None:
            return
        current = self.head
        while True:
            print(current.data, end=" <-> ")
            current = current.next
            if current == self.head:
                break
        print("(head)")

# Example usage
list = DoublyCircularLinkedList()
list.insert(3)
list.insert(2)
list.insert(1)

print("Original List:")
list.display()

list.delete_node(2)
print("After Deleting 2:")
list.display()
